### PR TITLE
Validate resource name

### PR DESCRIPTION
**Pull request by bandinib-amzn**
*Issue #, if available:*

*Description of changes:*
Validate resource name (role name, user name, tenant name, permission group name/ single permission name)

*Validation Criteria:*

- The resource name must contain from 2 to 50 characters.

- Valid characters are A-Z, a-z, 0-9, (_)underscore, (-) hyphen and unicode characters.

*Testing:*

> Role name

image

> User name

image

> Tenant name

image

> Permission

image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: public/apps/configuration/constants.tsx [hunk 1]
```diff
@@ -156,3 +156,7 @@ export enum ToolTipContent {
   DocumentLevelSecurity = 'Document-level security lets you restrict a role to a subset of documents in an index.',
   FieldLevelSecurity = 'Field-level security lets you control which document fields a user can see.',
 }
+
+// Name constraint: m is lower limit  and n is upper limit
+export const m = 2;
+export const n = 50;
```
Comment (yuxitang-amzn) [inline public/apps/configuration/constants.tsx:162]: `m` and `n` is not very meaningful. And for shared constant the name conversion is all upper case. I would suggest use sth. like `MIN_NUMBER_OF_CHARS_IN_RESOURCE_NAME`

File: public/apps/configuration/panels/internal-user-edit/internal-user-edit.tsx [hunk 2]
```diff
@@ -142,19 +145,23 @@ export function InternalUserEdit(props: InternalUserEditDeps) {
...
             <EuiFieldText
               value={userName}
               onChange={(e) => {
                 setUserName(e.target.value);
               }}
+              onBlur={() => {
+                checkForResourceNameErrors('user', userName);
+              }}
               disabled={props.action === 'edit'}
+              isInvalid={showErrors}
```
Comment (yuxitang-amzn) [inline public/apps/configuration/panels/internal-user-edit/internal-user-edit.tsx:164]: Shall we consider creating a shared component for it?

File: public/apps/configuration/utils/resource-validation-util.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  showErrors: boolean;
+  errors: string[];
+  checkForResourceNameErrors: (resourceType: string, resourceName: string) => void;
+}
+
+export function useErrorState(): UseErrorState {
+  const [showErrors, setShowErrors] = useState<boolean>(false);
+  const [errors, setErrors] = useState<string[]>([]);
+
+  const checkForResourceNameErrors = (resourceType: string, resourceName: string) => {
```
Comment (yuxitang-amzn) [inline public/apps/configuration/utils/resource-validation-util.tsx:39]: The naming is a bit confusing to me. `useErrorState` sounds very general, but it actually cares resource name specifically. My suggestion is move to a shared NameRow component that accepts a `setErrors` function. And for showErrors I think we can use errors.length instead.

File: public/apps/configuration/constants.tsx (hunk 1 above)
Comment (bandinib-amzn) [reply public/apps/configuration/constants.tsx]: Used `MIN_NUMBER_OF_CHARS_IN_RESOURCE_NAME` in place of `m` and `MAX_NUMBER_OF_CHARS_IN_RESOURCE_NAME` in place of `n`

File: public/apps/configuration/panels/internal-user-edit/internal-user-edit.tsx (hunk 2 above)
Comment (bandinib-amzn) [reply public/apps/configuration/panels/internal-user-edit/internal-user-edit.tsx]: Yes. Created shared component named `NameRow`

File: public/apps/configuration/utils/resource-validation-util.tsx (hunk 3 above)
Comment (bandinib-amzn) [reply public/apps/configuration/utils/resource-validation-util.tsx]: Removed `UseErrorState`. Added `setErrors` hook in `NameRow` component.

File: public/apps/configuration/utils/name-row.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  const validateName = () => {
+    setErrors(validateResourceName(props.resourceType, props.resourceName));
+    const errorMessages = validateResourceName(props.resourceType, props.resourceName);
+    if (errorMessages.length > 0) {
+      props.setIsFormValid(false);
+      setErrors(errorMessages);
+    } else {
+      props.setIsFormValid(true);
+      setErrors([]);
+    }
```
Comment (yuxitang-amzn) [inline public/apps/configuration/utils/name-row.tsx:43]: Is it equivalent to 

~~~
props.setIsFormValid(errorMessages.length > 0);
setErrors(errorMessages);
~~~

**Review by yuxitang-amzn — APPROVED** [review]
Approve with some minors.

File: public/apps/configuration/panels/internal-user-edit/internal-user-edit.tsx [hunk 5]
```diff
@@ -143,20 +144,15 @@ export function InternalUserEdit(props: InternalUserEditDeps) {
...
-            <EuiFieldText
-              value={userName}
-              onChange={(e) => {
-                setUserName(e.target.value);
-              }}
-              disabled={props.action === 'edit'}
-            />
-          </FormRow>
+            resourceName={userName}
+            resourceType={'user'}
```
[the remark this answers is not in the document]
Comment (bandinib-amzn) [reply public/apps/configuration/panels/internal-user-edit/internal-user-edit.tsx]: fixed.

File: public/apps/configuration/utils/name-row.tsx (hunk 4 above)
Comment (bandinib-amzn) [reply public/apps/configuration/utils/name-row.tsx]: right. refactored.

**Pull request merged**